elastic/kibana
# [Response Ops][Alerting] Ad hoc rule task runner (merging into feature branch)

Resolves https://github.com/elastic/kibana/issues/174358

Note that this merges into a feature branch

## Summary

Adds task runner for ad hoc runs of rules over an arbitrary time interval. This PR:
1. Updates the `BackfillClient` to create a task (`type:ad_hoc_run-backfill`) for each backfill job scheduled via the [schedule API](https://github.com/elastic/kibana/pull/176185)
2. Creates an `AdHocTaskRunner` to run these tasks. This task runner does the following:
  a. Loads and decrypts the `AdHocRun` saved object
  b. Determines which schedule entry from the saved object to run by looking for the next `PENDING` entry
  c. Uses the `runAt` time in the schedule as the `startedAt` time passed into the executors and to determine the time range returned by the `getTimeRange` function.
  d. Creates an `execute-backfill` event log doc that contains the ID of the `AdHocRun` saved object and the `runAt` and interval for the ad hoc run
  e. Updates the schedule entry in the `AdHocRun` saved object to reflect the outcome of the execution, either `success`, `timeout` or `error`.
3. Alerts created via the ad hoc task runner uses the following timestamps
  - `@timestamp` - this is set to the backfill `runAt` time
  - `kibana.alert.start` - this is set to the backfill `runAt` time
  - `kibana.alert.rule.execution.timestamp` - this is a new field that reflects the actual time the backfill was run. For real-time rule runs, this timestamp should match the `@timestamp` timestamp.
4. When all schedule entries for a backfill have been run, either successfully or unsuccessfully, the `ad_hoc_run-backfill` task will be deleted and the `AdHocRun` saved object will be deleted. We can use the event log entries to get the status of each backfill run, similar to what we do with action runs. We do this cleanup in order to clean up the tasks and saved objects. In the future, if we introduce ways to retry specific runs, we can change the logic to leave behind the task and/or SO where it makes sense.

## To Verify
- Index some documents with timestamps in the past.
- Create a detection rule that queries over the index and grab the ID of the rule. I typically choose a long schedule interval like `12h` so that the actual rule only runs once during testing.
- Use the schedule backfill API to schedule a backfill. Pick a long backfill time range so it doesn't run too fast

```
POST /internal/alerting/rules/backfill/_schedule
[
  {
    "rule_id": <ruleID>,
    "start": "2023-12-30T12:00:00.000Z",
    "end": "2024-01-01T12:00:00.000Z",
  }
]
```

- Verify that a task was created for this backfill. The task should use the same ID as the backfill
- Query for the `AdHocRun` SO in Dev Tools. You should be able to see the schedule entries changing status as the backfill executions are fulfilled.
- Verify that the appropriate event log docs are written for the backfills. There should be one `execute-backfill` event written for each schedule entry
- Verify that alerts are found and the timestamp fields for these alerts are populated correctly.
